Create user API

kashizui opened the pull request:
Implements the users API.
- Fetch and update the authenticated user at `/rest/user`.
- Fetch single users at `/rest/users/<id>`.
- List/search users at `/rest/users`
  - Supports `?filter[user_name]=name1,name2` and ?filter[email]=email1,email2`query params

Also adds unique constraints on username and email fields, and updates the migrate-users script to work with Django app backed by sqlite.

Contributes to #252

kashizui committed:
Build users REST API

kashizui commented on codalab/lib/jsonapi.py:
```diff
@@ -0,0 +1,21 @@
+    for key, value in query.items():
+        m = re.match(FIELDSET_KEY_PATTERN, key)
+        if m is None:
+            print key
```
delete

kashizui committed:
Fix migrate script to work with sqlite

kashizui commented:
@klopyrev @percyliang please review

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+    return UserSchema(only=fieldset).dump(request.user).data
+
 
+@route('/user', method=['PUT', 'PATCH'], apply=[JsonApiPlugin(), AuthenticatedPlugin()])
```
Why PUT?

kashizui replied:
You are right, I should not be using `PUT`.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+    try:
+        user_info, errors = UserSchema(strict=True).load(request.json, partial=True)
+    except ValidationError as e:
+        response.status = '403 Forbidden'
```
abort(httplib.FORBIDDEN, e.messages)

klopyrev replied:
Is 403 really the right error code? Read its description carefully: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

kashizui replied:
I agree, 403 is the wrong one to use. Which one would you suggest? I can't seem to find one that fits.. are we supposed to invent our own?

kashizui replied:
Oops should be `400 Bad Request`!

klopyrev replied:
Yes, thanks.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+    # Ensure that user name is not taken
+    if (user_info.get('user_name', request.user.user_name) != request.user.user_name and
+        local.model.user_exists(user_info['user_name'], None)):
+        response.status = '403 Forbidden'
```
You have wildly different error handling code. In one place you just set status, in another you raise a UsageError, in a third (below) you return an HTTPResponse with code 404. Please be consistent. Personally, I would do away with the JSON API errors return value, and simply use HTTP Error codes and body. See these two pull requests:

https://github.com/codalab/codalab-cli/pull/394
https://github.com/codalab/codalab-worksheets/pull/120

kashizui replied:
This is great, I will incorporate these patterns.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+class UserSchema(Schema):
+    id = fields.String(dump_only=True, attribute='user_id')
+    user_name = fields.String()
+    email = fields.Email(dump_only=True)
```
Why can't you change the email?

kashizui replied:
I disabled it for now until I've implemented the re-verification process (changing email should require verification of the new email address.)

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+    return UserSchema(only=fieldset).dump(user).data
+
+
+@route('/users/<id>', method=['PUT', 'PATCH'], apply=JsonApiPlugin())
```
Why PUT?

kashizui replied:
This `PUT` will also be removed.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,133 @@
+
+
+@route('/users/<id>', method=['PUT', 'PATCH'], apply=JsonApiPlugin())
+def update_user(id):
```
Why is this API needed?

kashizui replied:
Simply for convenience on the client side, but I'm happy to remove it if you think it's unnecessary.

klopyrev replied:
Yeah, it's just confusing.

klopyrev commented on codalab/server/json_api_plugin.py:
```diff
@@ -0,0 +1,110 @@
+def parse_includes(query):
+    """Includes currently not supported."""
+    if 'include' in query:
+        raise HTTPError('400 Bad Request')
```
abort(httplib.BAD_REQUEST), here an elsewhere.

klopyrev commented on codalab/server/json_api_plugin.py:
```diff
@@ -0,0 +1,110 @@
+"""
```
This JSON API stuff seems like overkill. Just my personal opinion.

kashizui replied:
I'll take a look at simplifying this.

klopyrev replied:
Evaluate whether you think it's worth it. Are you planning on migrating the rest of the code to use the JSON API? If so, you're welcome to do it. If not, then here it's just overkill that makes the code hard to read unless you're pretty familiar with the JSON API.

Take for example the fieldset stuff. Is it really necessary? What's wrong with just sending all the data. I can see it being useful with some worksheet stuff where reading the extra info requires an extra database call. Even then, at this point it's probably premature optimization.

kashizui replied:
You're right, I've decided against implementing anything more than necessary at this moment until we need it.

kashizui committed:
Strip down User API

kashizui commented:
@klopyrev Stripped down the implementation considerably, please take a look when you get the chance.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,109 @@
+    """
+    # Combine username and email filters
+    usernames = set(request.query.get('filter[user_name]', '').split(','))
+    usernames |= set(request.query.get('filter[email]', '').split(','))
```
If I query with filter[user_name]=blah this generates the set ['', 'blah']. The empty string for some reason matches all names. Thus, this code is broken.

Please test your code. Reviewers shouldn't generally have to do that for you.

kashizui replied:
I'm not able to reproduce this bug:
http://45.55.25.29:8000/rest/users?filter%5Buser_name%5D=sckoo
returns only the one matching user. 

But in fact there are bugs I didn't catch when I last tested it since `http://45.55.25.29:8000/rest/users` doesn't return any users.. I will do more thorough testing this time, sorry about that.

klopyrev replied:
I have an idea of why it's a bug. create-root-user.py creates a user with
no email and the empty string matches that. You shouldn't include the empty
string.

By the way, this API is not used...

On Fri, Apr 8, 2016 at 2:37 PM, Stephen Koo notifications@github.com
wrote:

> In codalab/rest/users.py
> https://github.com/codalab/codalab-cli/pull/409#discussion_r59092885:
> 
> > -    return UserSchema(only=PUBLIC_USER_FIELDS).dump(user).data
> >   +
> >   +
> >   +@get('/users', apply=[JsonApiPlugin()])
> >   +def fetch_users():
> > -    """Fetch list of users, filterable by username and email.
> >   +
> > -    Takes the following query parameters:
> > -        filter[user_name]=name1,name2,...
> > -        filter[email]=email1,email2,...
> >   +
> > -    Fetches all users that match any of these usernames or emails.
> > -    """
> > -    # Combine username and email filters
> > -    usernames = set(request.query.get('filter[user_name]', '').split(','))
> > -    usernames |= set(request.query.get('filter[email]', '').split(','))
> 
> I'm not able to reproduce this bug:
> http://45.55.25.29:8000/rest/users?filter%5Buser_name%5D=sckoo
> returns only the one matching user.
> 
> But in fact there are bugs I didn't catch when I last tested it since
> http://45.55.25.29:8000/rest/users doesn't return any users.. I will do
> more thorough testing this time, sorry about that.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/409/files/8b9d1d3be83dc4b6fbe8f6e2438f8f4a687d1e0f#r59092885

kashizui replied:
Ah the root user. I will make sure to discard the empty string.

This was part of Percy's requested features (as was the sparse fieldsets thing) but maybe as we discussed earlier, we can remove this since we should only implement things as we need them?

kashizui replied:
I certainly don't mind fewer things to maintain/test...

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,109 @@
+    """Fetch a single user."""
+    user = local.model.get_user(id)
+    if user is None:
+        abort(httplib.NOT_FOUND)
```
Can you add a string message? Otherwise, it says Unknown Error.

klopyrev commented on codalab/rest/users.py:
```diff
@@ -1,10 +1,109 @@
+    request.user = local.model.get_user(request.user.user_id)
+
+    # Return updated user
+    return fetch_authenticated_user()
```
This doesn't need to be a separate function.

klopyrev commented on codalab/server/json_api_plugin.py:
```diff
@@ -0,0 +1,15 @@
+
+    def apply(self, callback, route):
+        def wrapper(*args, **kwargs):
+            response.content_type = 'application/vnd.api+json'
```
This isn't necessary since by default if you return a dict the response content type is set to application/json by the JsonPlugin in bottle. In fact, this setting is override by that plugin.

kashizui committed:
Address user API comments

kashizui commented:
Addressed @klopyrev 's comments. @percyliang should we hold off on implementing a way to query for the users list until we actually need to use it? I implemented a basic one but as @klopyrev pointed out it's not being used right now (and nor is the `/rest/users/<id>` endpoint).

percyliang commented:
I think it's worth having the users list and being able to get info on
individual users.  It's useful for admin purposes, for example.

On Fri, Apr 8, 2016 at 3:50 PM, Stephen Koo notifications@github.com
wrote:

> Addressed @klopyrev https://github.com/klopyrev 's comments. @percyliang
> https://github.com/percyliang should we hold off on implementing a way
> to query for the users list until we actually need to use it? I implemented
> a basic one but as @klopyrev https://github.com/klopyrev pointed out
> it's not being used right now (and nor is the /rest/users/<id> endpoint).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/409#issuecomment-207639255

kashizui commented:
OK, then this is tested and ready for review @klopyrev .

kashizui committed:
Allow unverified users to edit profile

klopyrev commented:
LGTM
